### DISCUSSION TITLE
Vanilla JavaScript

### DISCUSSION THREAD

**Discussion by kobeaerts** [tailwindlabs/headlessui#344] [Ideas]
Well, the title says is all.
We have some wonderful bindings for React & Vue but sadly we live in an imperfect world with projects that don't use / need React or Vue.

Any chance we can get a Headless UI version for Vanilla JavaScript?

**Comment by adamwathan** [comment]
Hey! Can you share some examples of what you think the code would actually look like from a user-facing API perspective? Everyone kind of has different definitions of "vanilla JS".

My gut feeling is that this is not something we would ever pursue, but mostly because I can't for the life of me imagine what sort of APIs we could provide that would actually feel helpful and ergonomic. We are planning to build a version of Headless UI for Alpine.js which is what I think of as the vanilla version in my mind, as it's an extremely lightweight library that sits on top of regular old server-rendered HTML.

**Reply by royalicing** [reply]
Perhaps as Web Components?

**Reply by kobeaerts** [reply]
I was thinking about a "bring your own markup" way. Take the Switch element for example.
We would create the markup like this:

```
<button class="bg-teal-700
          relative inline-flex flex-shrink-0 h-[38px] w-[74px] border-2 border-transparent rounded-full cursor-pointer transition-colors ease-in-out duration-200 focus:outline-none focus-visible:ring-2  focus-visible:ring-white focus-visible:ring-opacity-75" id="headlessui-switch-1"><span class="sr-only">Use setting</span><span aria-hidden="true" class="translate-x-0
            pointer-events-none inline-block h-[34px] w-[34px] rounded-full bg-white shadow-lg transform ring-0 transition ease-in-out duration-200"></span></button>
```

And then couple the magic JavaScript likt this:
```
new Switch({
    element: document.querySelector('...'),
});
```

This JavaScript would add the needed aria-roles and event handles.

**Reply by adamwathan** [reply]
How would the classes be updated based on the state of the switch in that example?

**Reply by kobeaerts** [reply]
We could pass them in the JavaScript like this:

`new Switch({
    element: document.querySelector('...'),
    activeClass: 'bg-blue-700',
});`

or we could work with data-attributes:

`<button class="bg-teal-700" data-activated-class="bg-blue-700">....</button>`

**Reply by xt0rted** [reply]
The web components I started working on do it all through attributes like this

```html
<transition-container
  class="absolute inset-0 transition-opacity bg-gray-500 bg-opacity-75 pointer-events-auto motion-reduce:transition-none js-slide-over-background"

  enter-class="duration-500 ease-in-out"
  enter-from-class="opacity-0"
  enter-to-class="opacity-100"

  leave-class="duration-500 ease-in-out"
  leave-from-class="opacity-100"
  leave-to-class="opacity-0"

  style="display: none;"
  hidden
>
</transition-container>
```

They're built using https://github.github.io/catalyst/ (still migrating over) and have zero dependencies when distributed.

The repo's private right now but the current samples can be seen at https://xt0rted.github.io/web-components/

**Reply by pbgneff** [reply]
Shoot, I just saw this discussion after I already started [a new one](https://github.com/tailwindlabs/headlessui/discussions/471) and I have no idea how to close a discussion.

I work with vanilla JS a lot, and have integrated a bunch of the TailwindUI components (including the transitions) into the regular code. Everything is written with JS no HTML or JSX and what not. Would that be of use/interest?

**Reply by niels-numbers** [reply]
>  We are planning to build a version of Headless UI for Alpine.js

@adamwathan when do you plan to release first alpha for Alpine.js? Any chance to support this project?

**Reply by adamwathan** [reply]
Can't promise any timeline, haven't started on it yet. Planning to experiment with it after Alpine v3 is released and figure out if it's viable.

**Reply by suneetk92** [reply]
> 
> 
> Can't promise any timeline, haven't started on it yet. Planning to experiment with it after Alpine v3 is released and figure out if it's viable.

Have a look at svelte, litElement, or stenciljs libraries which can build into web components.

**Reply by johnpangalos** [reply]
You could make up your own html tags and give them attributes, then run a JS function and replace it. I can give that a shot for a demo at least if anyone is interested.

**Reply by johnpangalos** [reply]
Might be worth seeing if this is compatible with the existing elements https://github.com/vuejs/petite-vue

**Reply by johnpangalos** [reply]
I imagine that headless ui uses refs though which petite vue doesn’t support

**Reply by adamwathan** [reply]
Doesn't support render functions either, and every single one of our components is authored entirely as render functions. Writing a version of Headless UI for something like Alpine is going to be a pretty massive project that will require a lot of API innovation because the entire approach has to be _completely_ different.

Right now it's still a "would love to play with this one day" idea but unfortunately no actual concrete plans to break ground on it — we've got the next 9-12 months already booked with other projects that are more important for just being able to keep the lights on and pay the bills I'm afraid. I'd still love to do it but realistically with only 2 of us working on this stuff it doesn't feel right to give any false hope of us playing with it any time soon.

**Reply by johnpangalos** [reply]
Really nice of you to respond, I appreciate the answer. That's a fair point about the size of team and priorities. Good luck! I enjoy all the projects tailwind labs makes.

**Reply by cmer** [reply]
Have you made progress on this @xt0rted? It seems like an interesting approach. Any plan to open source?

> The repo's private right now but the current samples can be seen at https://xt0rted.github.io/web-components/

**Reply by xt0rted** [reply]
@cmer I had some health & personal issues come up and I haven't had time to work on this, or any open source work really. I was building these out as part of another project which is also on hold for now but I'd like to get back to it and open source it when things improve.

**Reply by cmer** [reply]
Thanks for the reply. I hope thinks are going for the best for you.

**Reply by marshalhayes** [reply]
I'm considering making a proof of concept with web components. I've never used them, and have only a little experience with headlessui, but nevertheless it'd be fun to work on.

Would anyone else find this useful? At the beginning, I'm sure it'd be more of an independent project until the API is solidified.

**Reply by cmer** [reply]
that'd be pretty interesting!

**Reply by marshalhayes** [reply]
I've come up with [marshalhayes/headless-webcomponents](https://github.com/marshalhayes/headless-webcomponents) so far. The code is admittedly kinda terrible at the moment, but I'm curious to see if the idea is there.

The component I started with is called `@hui/switch` and is a typical toggle/switch implemented as a custom web component. Put simply, all it does is add a click event that toggles `HuiSwitch.checked`. It's value is reflected as an attribute in the DOM, and can be initially set there as well.

```html
<hui-switch checked></hui-switch>
```

In order to support similar functionality to headlessui, such as adding/removing classes based on the state for example, there are two other attributes available (`checked-class`, and `unchecked-class`) both on `<hui-switch />` itself and any of its children.

```html
<hui-switch 
  class="relative inline-flex items-center h-6 rounded-full w-11 bg-gray-200"
  checked-class="bg-blue-600"
  unchecked-class="bg-gray-200"
>
   <span class="translate-x-1 inline-block w-4 h-4 transform bg-white rounded-full"
      checked-class="translate-x-6"
      unchecked-class="translate-x-1"
  >
  </span>
</hui-switch>
```

<details>
<summary>The above is equivalent to <a href="https://headlessui.dev/react/switch#basic-example">this example in the headlessui/react docs</a>.</summary>

```jsx
import { useState } from 'react'
import { Switch } from '@headlessui/react'

function MyToggle() {
  const [enabled, setEnabled] = useState(false)

  return (
    <Switch
      checked={enabled}
      onChange={setEnabled}
      className={`${
        enabled ? 'bg-blue-600' : 'bg-gray-200'
      } relative inline-flex items-center h-6 rounded-full w-11`}
    >
      <span className="sr-only">Enable notifications</span>
      <span
        className={`${
          enabled ? 'translate-x-6' : 'translate-x-1'
        } inline-block w-4 h-4 transform bg-white rounded-full`}
      />
    </Switch>
  )
}
```

<p>The <code>onChange</code> handler isn't required for <code>&lt;hui-switch /&gt;</code>; however, if you need to perform actions when the switch is toggled, listen for the <code>hui.switch.toggle</code> event. It supports cancellation but doesn't bubble past the component itself.</p>
</details>

This approach is limited to only classes, so doing something more complex - especially with components more feature-rich than this simple example - might be difficult, but I'm curious to see what people think.

**Reply by marshalhayes** [reply]
The only reason I'm even considering an approach like this is because, like @kobeaerts said, not every application uses or needs React or Vue. An AlpineJS option would be interesting, but I'm toying with the idea of a zero-dependency custom web component library. No framework needed, just drop the script tag to load the component and use it just like a native element. 

No idea if that's possible (or wise) though 🤷🏻‍♂️

**Reply by inwardmovement** [reply]
@adamwathan do you still plan to make an Alpine.js version?

**Reply by adamwathan** [reply]
No plans for it I’m afraid, don’t have the resources for it.

**Reply by maitijoanne** [reply]
After a year of working with my team on integrating the [French government's design system](https://gouvfr.atlassian.net/wiki/spaces/DB/pages/217251916/l+ments+d+interface) as web components, I am convinced that web components are the ultimate solution for unstyled and fully accessible ui components, you can use everywhere. It's a perfect match with headlessui/tailwindcss.
We use [StencilJs](https://stenciljs.com/). Apart many optimizations, it generates component wrappers for vue, react and angular.
The headlessui's code is a bit hard for me, but I can try to make a POC with StencilJs, if you are interested.
